### DISCUSSION TITLE
Special Lazy<> support in AutoData for delayed fixture creation

### DISCUSSION THREAD

**Discussion by rcdailey** [AutoFixture/AutoFixture#1285] [Q&A]
I am using `AutoData` with `AutoMoqCustomization` to auto-inject constructor parameters for my SUT. One particular SUT has some business logic that happens in a constructor which requires the mocked & frozen injected dependencies to be set up prior to the constructor executing. My workaround right now is as follows (using Moq + NUnit3 + Autofixture):

```cs
[Test,
 InlineAutoMockData(DbPermissionsMode.ElevatedDefault, true),
 InlineAutoMockData(DbPermissionsMode.ReducedNoViewServerState, false)]
public void Enumerator_mappings_are_correct(
    DbPermissionsMode permissionMode, bool expectedBool,
    [Frozen] Mock<IProductConfigProvider> configProvider,
    IFixture fixture)
{
    configProvider.Setup(x => x.GetDatabasePermissionsMode(It.IsAny<bool>()))
                  .Returns(permissionMode);

    var vm = fixture.Build<CredentialsControlViewModel>()
                    .OmitAutoProperties()
                    .Create();

    vm.UseElevatedPermissions.Should().Be(expectedBool);
}
```

In this test, `CredentialsControlViewModel` invokes `IProductConfigProvider.GetDatabasePermissionsMode()` in its constructor. This workaround allows me to perform setup on the `configProvider` instance prior to constructing the `vm` variable using the injected fixture.

What would be more ideal to clean this up is something like this, which in my mind should be functionally the same:

```cs
[Test,
 InlineAutoMockData(DbPermissionsMode.ElevatedDefault, true),
 InlineAutoMockData(DbPermissionsMode.ReducedNoViewServerState, false)]
public void Enumerator_mappings_are_correct(
    DbPermissionsMode permissionMode, bool expectedBool,
    [Frozen] Mock<IProductConfigProvider> configProvider,
    [NoAutoProperties] Lazy<CredentialsControlViewModel> vm)
{
    configProvider.Setup(x => x.GetDatabasePermissionsMode(It.IsAny<bool>()))
                  .Returns(permissionMode);

    vm.Value.UseElevatedPermissions.Should().Be(expectedBool);
}
```

Here, I no longer use the `IFixture` directly. Instead, I want `Lazy<>` to behave as follows:

1. Internally create the given type (`CredentialsControlViewModel`) exactly as it would have been created if I instead passed in the parameter to the test as `CredentialsControlViewModel vm`.
2. Propagate all of the attached attributes to the type created by Lazy's factory function (append all of the customizations / specimen builders)
3. In all reasonable aspects, act as if I had done `[NoAutoProperties] CredentialsControlViewModel vm` directly, just within a factory function provided to `Lazy<>` instead of done immediately.

I see that there's already a `LazyRelay`, but it doesn't do this, I think. At least, I get an exception from Lazy saying it couldn't find a default constructor.

Could someone recommend a way to customize AutoFixture to get this behavior for Lazy?

**Comment by aivascu** [comment]
@rcdailey if I understand correctly your question, the issue you're having is that the NoAutoProperties customization is not applying to the `CredentialsControlViewModel` but rather to the `Lazy<CredentialsControlViewModel>`. 

The answer is that unfortunately the `NoAutoPropertiesAttribute` applies the customization to the type of the parameter (i.e. `Lazy<T>`).

What you could do is create your custom `NoAutoPropertiesAttribute` that would take into account if it is applied to a Lazy parameter.

```cs
public sealed class NoLazyAutoPropertiesAttribute : CustomizeAttribute
{
    public override ICustomization GetCustomization(ParameterInfo parameter)
    {
        if (parameter == null) throw new ArgumentNullException(nameof(parameter));

        var targetType = parameter.ParameterType;

        return targetType.IsGenericType && targetType.GetGenericTypeDefinition() == typeof(Lazy<>)
            ? new NoAutoPropertiesCustomization(targetType.GetGenericArguments().First())
            : new NoAutoPropertiesCustomization(targetType);
    }
}
```

**Reply by rcdailey** [reply]
I'm sorry, that's not the issue. The attributes applying to the underlying type for `Lazy` was just one aspect of my proposal for how the factory method should function.

Ultimately what I'm asking for is deferred instantiation of `CredentialsControlViewModel` by AutoFixture when it is wrapped in `Lazy<>`. This is so that I can set up mocks prior to construction of the SUT (which again is `CredentialsControlViewModel` in this case).

**Reply by aivascu** [reply]
@rcdailey I don't think this is possible. The only solution is to create the SUT after you've set up the mocks. Like you did in the first example of the question. 
As I see it there is a bigger issue here, and that is that you have business logic in the constructor. Constructors should be free of any kind of business logic. Having business logic in the constructor is considered bad practice for the exact reason you struggle with the unit test. There is no way to exercise control over this logic.
Most MVVM frameworks have some way of executing code right after the view model creation. You should check what is the recommended approach in your case.

**Reply by rcdailey** [reply]
In my case I'm just setting up some observable subscriptions in my constructor, which seems to be the standard practice. I feel like that is the exception to the rule.

In my mind I definitely feel like what I'm doing is possible. I just need to find out how AutoFixture customization works. There's zero documentation (that I could find, at least).

**Reply by aivascu** [reply]
@rcdailey I just published a [draft article](https://autofixture.github.io/docs/fixture-customization/) on Fixture customization. Feel free to add a issue on the documentation website's [repository](https://github.com/AutoFixture/AutoFixture.github.io/issues) to tell me what information is missing from the article.

**Comment by aivascu** [comment]
@rcdailey I think there might be a way to defer the viewmodel creation, but it would require you to move the viewmodel customization inside the AutoData attribute.
First you would need to create a customization that would inject a factory for the viewmodel into the fixture, and would also instruct the fixture to omit autoproperties on the viewmodel.
```cs
internal class MyCustomization : ICustomization
{
    public void Customize(IFixture fixture)
    {
        fixture.Customize<CredentialsControlViewModel>(c => c.OmitAutoProperties());
        fixture.Inject<Func<CredentialsControlViewModel>>(() => fixture.Create<CredentialsControlViewModel>());
    }
}
```
Then you could create a custom attribute that would apply the customization.
```cs
public class InlineDomainData: InlineAutoDataAttribute
{
    public InlineDomainData(params object[] args)
        : base(() => new Fixture().Customize(
            new CompositeCustomization(
                new AutoMoqCustomization { ConfigureMembers = true },
                new MyCustomization())), args)
    {
    }
}
```
Finally you should be able to run the following test
```cs
[Test,
 InlineDomainData(DbPermissionsMode.ElevatedDefault, true),
 InlineDomainData(DbPermissionsMode.ReducedNoViewServerState, false)]
public void Enumerator_mappings_are_correct(
    DbPermissionsMode permissionMode, bool expectedBool,
    [Frozen] Mock<IProductConfigProvider> configProvider,
    Func<CredentialsControlViewModel> factory)
{
    configProvider.Setup(x => x.GetDatabasePermissionsMode(It.IsAny<bool>()))
                  .Returns(permissionMode);
    var sut = factory();

    sut .UseElevatedPermissions.Should().Be(expectedBool);
}
```
It should be possible to use `Lazy<>` if you wish, you'd just have to inject a `Lazy<>` that uses the same lambda as the `Func <>` above.

**Reply by rcdailey** [reply]
Thank you for this. I feel like this is really close. I'd somehow need `MyCustomization` to remain an open generic type so that it registers any and all types that it sees inside `Lazy<>`. I don't want to keep coming back to this class and adding registrations. I'm happy to try to figure that out with you.

**Comment by Noggog** [comment]
This is how I got around my issues.  Same problem with RxUI VMs needing their injections prepped before they're created

```
public class LazyBuilder : ISpecimenBuilder
{
    public object Create(object request, ISpecimenContext context)
    {
        if (request is Type type
            && IsLazyType(type))
        {
            return GetCreateMethod()
                .MakeGenericMethod(type.GenericTypeArguments[0])
                .Invoke(this, new object[] {context, request})!;
        }
        
        return new NoSpecimen();
    }

    public static bool IsLazyType(Type t)
    {
        return t.GenericTypeArguments.Length == 1
               && t.Name.StartsWith("Lazy");
    }

    public static MethodInfo GetCreateMethod()
    {
        return typeof(LazyBuilder)
            .GetMethod("CreateLazy", BindingFlags.Static | BindingFlags.Public)!;
    }

    public static Lazy<T> CreateLazy<T>(
        ISpecimenContext context,
        object request)
    {
        return new Lazy<T>(() => (T)context.Create<T>());
    }
}
```
